contracts/l2/L2DAITokenBridge.sol

```diff
@@ -0,0 +1,180 @@
+  address public l1Token;
+
+  constructor(address _l2CrossDomainMessenger, address _l2Token) public OVM_CrossDomainEnabled(_l2CrossDomainMessenger) {
+    wards[msg.sender] = 1;
+    emit Rely(msg.sender);
+
+    l2Token = _l2Token;
+  }
+
+  modifier onlyAfterInit() {
+    require(l1DAITokenBridge != address(0), 'L2DAITokenBridge/not-init'); 
+    _;
+  }
+
+  function init(address _l1Gateway, address _l1Token) external auth {
+    require(l1DAITokenBridge == address(0), 'L2DAITokenBridge/already-init'); 
+
+    l1DAITokenBridge = _l1Gateway;
+    l1Token = _l1Token;
+  }
+
+  function close() external auth {
+    isOpen = 0;
+  }
+
+  function withdraw(
+        address _l2Token,
+        uint256 _amount,
+        uint32 _l1Gas,
+        bytes calldata _data
+    )
+        external
+        override
+        virtual
+        onlyAfterInit()
+    {
+        require(_l2Token == l2Token, "L2DAITokenBridge/token-not-dai");
+
+        _initiateWithdrawal(
+            msg.sender,
+            msg.sender,
+            _amount,
+            _l1Gas,
+            _data
+        );
+    }
+
+    function withdrawTo(
+        address _l2Token,
+        address _to,
+        uint256 _amount,
+        uint32 _l1Gas,
+        bytes calldata _data
+    )
+        external
+        override
+        virtual
+        onlyAfterInit()
+    {
+        require(_l2Token == l2Token, "L2DAITokenBridge/token-not-dai");
+
+        _initiateWithdrawal(
+            msg.sender,
+            _to,
+            _amount,
+            _l1Gas,
+            _data
+        );
+    }
+
+  // When a withdrawal is initiated, we burn the withdrawer's funds to prevent subsequent L2 usage.
+  function _initiateWithdrawal(
+        address _from,
+        address _to,
+        uint256 _amount,
+        uint32 _l1Gas,
+        bytes calldata _data
+    )
+        internal
+    {
+    // do not allow initiaitng new xchain messages if bridge is closed
+    require(isOpen == 1, 'L2DAITokenBridge/closed');
+
+    Mintable(l2Token).burn(msg.sender, _amount);
+
+    bytes memory message = abi.encodeWithSelector(
+        iOVM_L1ERC20Bridge.finalizeERC20Withdrawal.selector,
+        l1Token,
+        l2Token,
+        _from,
+        _to,
+        _amount,
+        _data
+    );
+
+    // Send message up to L1 bridge
+    sendCrossDomainMessage(
+        l1DAITokenBridge,
+        _l1Gas,
+        message
+    );
+
+    emit WithdrawalInitiated(l1Token, l2Token, msg.sender, _to, _amount, _data);
+  }
+
+  // When a deposit is finalized, we credit the account on L2 with the same amount of tokens.
+  function finalizeDeposit(
+        address _l1Token,
+        address _l2Token,
+        address _from,
+        address _to,
+        uint256 _amount,
+        bytes calldata _data
+    ) 
+    external
+    override
+    virtual
+    onlyAfterInit()
+    onlyFromCrossDomainAccount(l1DAITokenBridge)
+  {
+    require(_l1Token == l1Token && _l2Token == l2Token, "L2DAITokenBridge/token-not-dai");
```

Comment:
Okay I can't actually think of any cases where this might fail (maybe during upgrades?) but if it does fail then the deposited DAI in the L1 bridge will be locked. To be safe you could do what we do and instead of reverting just immediately initiate a withdrawal https://github.com/ethereum-optimism/optimism/blob/4a858d59b923a1ab2c2433b646e48b1f52595030/packages/contracts/contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2StandardBridge.sol#L184

Reply:
mhm I thought about this too. But I can't think of any situation when this can happen other than misconfiguration. But then no transfer will work. 

If we upgrade then there will be a new token bridge deployed using same l1/l2 tokens so this should be fine.